org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/context/InstalledBundleExtension.java

```diff
@@ -0,0 +1,157 @@
+			.getAllInstances()) {
+			final Class<?> testClass = instance.getClass();
+
+			List<Field> fields = findAnnotatedNonStaticFields(testClass, InjectInstalledBundle.class);
+
+			fields.forEach(field -> {
+				assertFieldIsOfType(field, Bundle.class, InjectInstalledBundle.class, ExtensionConfigurationException::new);
+				InjectInstalledBundle injectBundle = field.getAnnotation(InjectInstalledBundle.class);
+				setField(field, instance, getBundle(extensionContext, injectBundle));
+			});
+		}
+	}
+
+	/**
+	 * Resolve {@link Parameter} annotated with
+	 * {@link InjectBundleContext @BundleContextParameter} OR
+	 * {@link InjectBundleInstaller @InstallBundleParameter} in the supplied
+	 * {@link ParameterContext}.
+	 */
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		Parameter parameter = parameterContext.getParameter();
+		Class<?> parameterType = parameter.getType();
+
+		if (parameterContext.isAnnotated(InjectInstalledBundle.class)) {
+			assertParameterIsOfType(parameterType, Bundle.class, InjectInstalledBundle.class, ParameterResolutionException::new);
+			InjectInstalledBundle injectBundle = parameter.getAnnotation(InjectInstalledBundle.class);
+			return getBundle(extensionContext, injectBundle);
+		}
+
+		throw new ExtensionConfigurationException("No parameter types known to BundleContextExtension were found");
+	}
+
+	/**
+	 * Determine if the {@link Parameter} in the supplied
+	 * {@link ParameterContext} is annotated with
+	 * {@link InjectBundleContext @BundleContextParameter} OR
+	 * {@link InjectBundleInstaller @InstallBundleParameter}.
+	 */
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		boolean annotatedBundleParameter = parameterContext.isAnnotated(InjectInstalledBundle.class);
+
+		if (annotatedBundleParameter && (parameterContext.getDeclaringExecutable() instanceof Constructor)) {
+			throw new ParameterResolutionException(
+				"BundleContextExtension does not support parameter injection on constructors");
+		}
+		return annotatedBundleParameter;
+	}
+
+	private Bundle getBundle(ExtensionContext extensionContext, InjectInstalledBundle injectBundle) {
+
+		return installBundleOf(injectBundle, extensionContext);
+	}
+
+	private Bundle installBundleOf(InjectInstalledBundle injectBundle, ExtensionContext extensionContext) {
+
+		try {
+			BundleContext bc = BundleContextExtension.getBundleContext(extensionContext);
+			BundleInstaller ib = BundleContextExtension.getInstallBundle(extensionContext);
+
+			String spec = injectBundle.value();
+			if (spec.startsWith("http:") || spec.startsWith("https:") || spec.startsWith("file:")) {
+				return ib.installBundle(new URL(injectBundle.value()), injectBundle.start());
+			} else {
+				return ib.installBundle(BundleInstaller.EmbeddedLocation.of(bc, spec), injectBundle.start());
+			}
+		} catch (MalformedURLException e) {
```

Comment:
What will `EmbeddedLocation.of()` do if the spec is malformed? We're expecting our users to play nice but we can't enforce that with compile-time type safety as it is a plain `String`. A quick glance suggests that it will throw `IllegalArgumentException`. Suggest 1. Adding a test case for this, and then 2. adding `| IllegalArgumentException` to the type of the catch clause.

Reply:
added a test